### ISSUE TITLE
Add cpu pinning to nodes configuration

### ISSUE THREAD

**Issue by hedibouattour** [kubernetes-sigs/kind#3138]
**What would you like to be added**:
The possibility to pin nodes to cpus in kind clusters.
**Why is this needed**:
For large clusters (scale testing purposes), nodes overload the cpu. So we need to pin kind nodes to sets of specific cpus during their creation. 
As an example, on a 96 thread machine, kind clusters with Calico start to show instability above 30 nodes, this patch enables stability with 60 kind workers, by pinning every node to a cpu.
Docker allows this using the `--cpuset-cpus` flag. The easiest way is to add that to the node configuration like in https://github.com/kubernetes-sigs/kind/pull/3131.

**Comment by BenTheElder** [comment]
> As an example, on a 96 thread machine, kind clusters with Calico start to show instability above 30 nodes, this patch enables stability with 60 kind workers, by pinning every node to a cpu.


Wow! Can you elaborate more on the use case?

While kind technically supports multi-node, this is for https://kind.sigs.k8s.io/docs/contributing/project-scope/#p0-support-testing-kubernetes
Where we need to test some rolling node behavior.

It's not intended as a reasonable solution for performance testing.

> Docker allows this using the --cpuset-cpus flag

Mmm, is there a comparable flag in podman etc? We've tried to minimize coupling to docker going forward so we can attempt to meet demands to support other tools (e.g. also nerdctl, ignite...)

What's the expected user experience here?  Users need to select CPUs for each node?

**Comment by BenTheElder** [comment]
Is this correctly transitively applied to the "nested" containers? Under what environment have you tested this? (It's been a while since looking at this part of the system and we have podman/docker, runc/crun, rootless/rootful, ...)

**Comment by BenTheElder** [comment]
Kubernetes usually scale tests with multiple VMs or physical machines, this is a new one 😅

**Comment by dims** [comment]
xref: https://github.com/kubernetes-sigs/kind/pull/3131

**Comment by BenTheElder** [comment]
We can't move forward with something like this without digging in deeper to the interactions with the various container environments etc to make sure we're not introducing a new footgun.

See also: #3169 #2999

**Comment by sknat** [comment]
Right, I agree on the fact that we probably don't have a full picture of the consequences the 
use of such an option would have, and yes probably using it would break some of the usual guarantees of kind. 

Would having an more generic option e.g. `extraContainerRuntimeArgs` then be a viable alternative ?
This would allow building more complex behaviors like this one, while making it clear that the
option is highly dependent on the runtime implementation, and that consequences are also up 
to the caller to handle.

**Comment by BenTheElder** [comment]
> Would having a more generic option e.g. extraContainerRuntimeArgs then be a viable alternative ?

no. It's an implementation detail that we exec docker and it will be a nightmare to support this in the future. Consider if kind needs to start setting an argument that previously a user set or if they conflict or if we opt to switch to a client library.

**Comment by BenTheElder** [comment]
We can support pinning nodes but we need more investigation and some thought about the abstraction before committing. Typically we try to borrow from CRI if possible.

we've been bitten many times by thin abstractions and compat nightmares. See e.g. #2875 where we're in a bind because we didn't shim this like kOps did.
